src/runtime/slot-composer.ts

```diff
@@ -187,13 +187,8 @@ export class SlotComposer {
     return this._contexts;
   }
 
-  dispose(arc: Arc) {
-    this.consumers.forEach(consumer => {
-      // At this point there should be a single Arc per SlotComposer.
-      // TODO: Fix disposal once multi-arc SlotComposer is possible.
-      assert(consumer.arc === arc);
-      consumer.dispose();
-    });
+  dispose() {
+    this.consumers.forEach(consumer => consumer.dispose());
```

Comment:
is it correct that slot-composer handles a single outer arc and all its inner arcs? maybe then we should keep the 1:1 relation between arc and slot-composer, and create a new slot-composer for each inner arc with only its hosted slots in it?

i'm also curious to see how all this will work with a nested inner arc, could you maybe create a unit test for that?

Reply:
I am concerned about a case where a take a slot from the inner arc and make it available in the outer arc - which I think we're very likely to make possible. When that happens, it will no longer be a tree structure, it will be a graph (or something like that), and I think it's much cleaner to have all of that in a single SlotComposer. That what I wanted to do with the previous patch.

By a nested inner arc do you mean an inner arc of an inner arc? That would be fun to test - will work on it now :)

Reply:
inner arc providing a slot for an outer arc? interesting. do you have an example use case in mind, i'm just curious?
from the arc(s) : slot-composer(s) perspective - after disposing one arc, we end up with other arcs that use the same slot composer holding on to an invalid (disposed) slot composer. I'm not saying we should resolve it now, but at least have an idea of a solution. even if it's a graph, there seems to always be a top-level outer arc, isn't there? maybe worth filing an issue to have a discussion about this?

yes, by nested inner arc i did mean an inner arc of an inner arc. thanks, i'm looking forward to that test :)

Reply:
I wrote the test for recursive inner arcs... and they don't work. I'll need to make some changes in the SlotComposer, but I would like to leave that to the next PR.

If you think of inner arc as an implementation detail of a particle, then exposing inner arc slots to the outer arc makes sense. Some particle may create an inner arc with some recipe structure, and expose some of the leaf slots of the inner arc back to the outer arc.

I think the model of a single SlotComposer composing all arcs visible at once makes sense, for now this is an outer arc with all its inner arcs.